NonZero constructor now returns Option

https://github.com/rust-lang/rust/pull/42959

```
error[E0308]: mismatched types
    --> src/de/impls.rs:1583:16
     |
1583 |             Ok(NonZero::new(value))
     |                ^^^^^^^^^^^^^^^^^^^ expected struct `core::nonzero::NonZero`, found enum `core::option::Option`
     |
     = note: expected type `core::nonzero::NonZero<T>`
                found type `core::option::Option<core::nonzero::NonZero<_>>`
     = help: here are some functions which might fulfill your needs:
             - .unwrap()
```

--- a/serde/src/de/impls.rs
+++ b/serde/src/de/impls.rs
@@ -1573,14 +1573,9 @@ where
         D: Deserializer<'de>,
     {
         let value = try!(Deserialize::deserialize(deserializer));
-        unsafe {
-            let ptr = &value as *const T as *const u8;
-            if slice::from_raw_parts(ptr, mem::size_of::<T>()).iter().all(|&b| b == 0) {
-                return Err(Error::custom("expected a non-zero value"));
-            }
-            // Waiting for a safe way to construct NonZero<T>:
-            // https://github.com/rust-lang/rust/issues/27730#issuecomment-269726075
-            Ok(NonZero::new(value))
+        match NonZero::new(value) {
+            Some(nonzero) => Ok(nonzero),
+            None => Err(Error::custom("expected a non-zero value")),
         }
     }
 }

